packages/tenant-discovery-react/src/plugin.tsx

```diff
@@ -0,0 +1,235 @@
+import { EmailPasswordPreBuiltUI } from "supertokens-auth-react/recipe/emailpassword/prebuiltui";
+
+import { getApi } from "./api";
+import { PLUGIN_ID, API_PATH } from "./constants";
+import { enableDebugLogs } from "./logger";
+import { SelectTenantPage } from "./pages";
+import { getOverrideableTenantFunctionImplementation } from "./recipeImplementation";
+import { defaultTranslationsTenantDiscovery } from "./translations";
+import {
+  OverrideableTenantFunctionImplementation,
+  SuperTokensPluginTenantDiscoveryPluginConfig,
+  SuperTokensPluginTenantDiscoveryPluginNormalisedConfig,
+  TranslationKeys,
+} from "./types";
+import { hidePasswordInput, parseTenantId, populateEmailFromUrl, updateSignInSubmitBtn } from "./util";
+
+const { usePluginContext, setContext } = buildContext<{
+  plugins: SuperTokensPublicPlugin[];
+  sdkVersion: string;
+  appConfig: SuperTokensPublicConfig;
+  pluginConfig: SuperTokensPluginTenantDiscoveryPluginNormalisedConfig;
+  querier: ReturnType<typeof getQuerier>;
+  api: ReturnType<typeof getApi>;
+  t: (key: TranslationKeys) => string;
+}>();
+export { usePluginContext };
+
+export const init = createPluginInitFunction<
+  SuperTokensPlugin,
+  SuperTokensPluginTenantDiscoveryPluginConfig,
+  OverrideableTenantFunctionImplementation,
+  SuperTokensPluginTenantDiscoveryPluginNormalisedConfig
+>(
+  (pluginConfig, implementation) => {
+    let apiBasePath: string;
+    let translations: ReturnType<typeof getTranslationFunction<TranslationKeys>>;
+    return {
+      id: PLUGIN_ID,
+      init: async (config, plugins, sdkVersion) => {
+        if (config.enableDebugLogs) {
+          enableDebugLogs();
+        }
+
+        const querier = getQuerier(new URL(API_PATH, config.appInfo.apiDomain.getAsStringDangerous()).toString());
+        const api = getApi(querier);
+
+        // Set up the usePlugin hook
+        apiBasePath = new URL(API_PATH, config.appInfo.apiDomain.getAsStringDangerous()).toString();
+        translations = getTranslationFunction<TranslationKeys>(defaultTranslationsTenantDiscovery);
+
+        setContext({
+          plugins,
+          sdkVersion,
+          appConfig: config,
+          pluginConfig,
+          querier,
+          api,
+          t: translations,
+        });
+
+        // Check if the url matches any tenantId and accordingly
+        // set the tenantId in the url and refresh the page.
+        const tenantId = await implementation.determineTenantFromURL();
+        if (tenantId) {
+          implementation.setTenantId(tenantId);
+          return;
+        }
+
+        // If the tenant was not determined from the URL and
+        // tenant selector is enabled, we will redirect to the
+        // select tenant page.
+        const { shouldShowSelector } = parseTenantId();
+        if (shouldShowSelector && pluginConfig.showTenantSelector && window.location.pathname !== "/select-tenant") {
+          window.location.href = "/select-tenant";
+        }
```

Comment:
We only want to do this on the auth page, right? Why do we redirect from the init? How do we handle public pages?